File: Trees_Graphs/go/bst.go
package main
import (
  "types"
  "fmt"
)


func validateBST(leftNode *types.TreeNode, middle *types.TreeNode, rightNode *types.TreeNode) bool {
  var left, right bool

  if leftNode == nil {
    if middle.Value > rightNode.Value {
      return false
    } else {
      return true
    }
  } else if rightNode == nil {
    if middle.Value <= leftNode.Value {
      return false
    } else {
      return true
    }
  } else {
    if leftNode.Value > middle.Value || rightNode.Value <= middle.Value {
      return false
    }
  }

  if leftNode.LeftChild == nil && leftNode.RightChild == nil {
    left = true
  } else {
    left = validateBST(leftNode.LeftChild, leftNode, leftNode.RightChild)
  }

  if rightNode.LeftChild == nil && rightNode.RightChild == nil {
    right = true
  } else {
    right = validateBST(rightNode.LeftChild, rightNode, rightNode.RightChild)
  }

  return left && right
}


func writeToArray(array *[]types.TreeNode, node *types.TreeNode) {
  if node.LeftChild != nil {
    writeToArray(array, node.LeftChild)
  }
  *array = append(*array, *node)

  if node.RightChild != nil {
    writeToArray(array, node.RightChild)
  }
}


func validateBST2(node *types.TreeNode) bool {
  var array []types.TreeNode
  writeToArray(&array, node.LeftChild)
  array = append(array, *node)
  writeToArray(&array, node.RightChild)

  for i := 0; i < len(array) - 1; i++ {
    if array[i].Value > array[i+1].Value {
      return false
    } else if array[i].Value == array[i+1].Value && array[i].RightChild != nil {
      if *array[i].RightChild == array[i+1] {
        return false
      }
    }
  }

  return true
}


func main() {
  array := []int{1, 2, 4, 4, 5, 6, 7, 8}
  root := types.CreateTree(array)
  types.PrintTreeNodes(root)

  fmt.Println(validateBST(root.LeftChild, root, root.RightChild)) // wrong
  fmt.Println(validateBST2(root))
}

File: Trees_Graphs/python/bst.py
from Tree import *


def write_to_array(array, node):
    if node.left_child is not None:
        write_to_array(array, node.left_child)

    array.append(node)

    if node.right_child is not None:
        write_to_array(array, node.right_child)


def validate_BST(node):
    array = []
    write_to_array(array, node.left_child)
    array.append(node)
    write_to_array(array, node.right_child)

    for i in range(len(array) - 1):
        if array[i].value > array[i + 1].value:
            return False
        elif array[i].value == array[i + 1].value and node.right_child is not None:
            return False

    return True


def main():
    array = list(range(7))
    array = [1, 2, 4, 4, 5, 6, 7, 8]
    tree = Tree(array)
    tree.print_tree()

    print(validate_BST(tree.root))


if __name__ == '__main__':
    main()
